fsw/src/crypto_props.c

```diff
@@ -0,0 +1,351 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include "stdbool.h"
+#include <string.h>
+#include "crypto_props.h"
+
+crypto_config_list* p_self = NULL; 
+/*=============================================================================
+Private Declarations
+==============================================================================*/
+int crypto_config_props_set_property_value_helper(char* key, char* value, crypto_config_list** props);
+char* crypto_config_props_get_property_value_helper(char* key, crypto_config_list** props);
+int crypto_config_props_add_property_helper(char* key, char* value,crypto_config_list** pp_self); 
+int crypto_config_props_remove_property_helper(char* key,crypto_config_list** pp_self); 
+int safe_copy(char* dest, char* src);
+/*=============================================================================
+ * Implementations
+==============================================================================*/
+crypto_config_list* crypto_config_alloc(int max_key_size, int max_value_size)
+{
+    if (NULL==p_self)
+    {
+        p_self = (crypto_config_list*) calloc(1,sizeof(crypto_config_list));
+        p_self->KEY_SIZE_IN_BYTES = max_key_size; 
+        p_self->VALUE_SIZE_IN_BYTES = max_value_size; 
+        
+    }
+    return p_self; 
+}
+
+void crypto_config_free(crypto_config_list** pp_self)
+{
+    if (NULL!=pp_self && NULL!=(*pp_self))
+    {
+        crypto_config_node* p_current = (*pp_self)->p_head; 
+        crypto_config_node* p_next = NULL; 
+        //iterate and free the nodes 
+        while (NULL!=p_current)
+        {
+            p_next = p_current->p_next; 
+            free(p_current);
+            p_current = p_next; 
+           
+        }
+        //free the super structure 
+        free (*pp_self);
+    }
+}
+
+int crypto_config_props_add_property(char* key, char* value)
+{
+    int success = 0; 
+    //Note: since this is a linked list there is no efficient way to guard against
+    //adding the duplicate keys therefore if the key exists just update it using
+    //crypto_config_props_set_property_value_helper else add in a new key,value. 
+    success = crypto_config_props_set_property_value_helper(key,value,&p_self);
+    if (success)
+    {
+        return success; 
+    }
+    else
+    {
+        return crypto_config_props_add_property_helper(key,value,&p_self);
+    }
+}
+int crypto_config_props_add_property_helper(char* key, char* value, crypto_config_list** pp_self)
+{
+    int success = 0; 
+    if (NULL!=key && strlen(key)> 0 && NULL!=value && NULL!=pp_self && NULL!=(*pp_self))
+    {
+       
+        //is the head NULL?  
+        if (NULL==(*pp_self)->p_head)
+        {
+            //if head is NULL then allocate memory for it 
+            (*pp_self)->p_head = calloc(1,sizeof(crypto_config_node)); 
+            (*pp_self)->p_head->key = calloc((*pp_self)->KEY_SIZE_IN_BYTES,sizeof(char)); 
+            (*pp_self)->p_head->value = calloc((*pp_self)->VALUE_SIZE_IN_BYTES,sizeof(char));
```

Comment:
key & value are passed in, and you already know their length... We do not need to allocate memory for KEY_SIZE_IN_BYTES, we can just allocate for strlen(key).

Reply:
Also, since this should be updated to not waste space, we should use a malloc here instead of a calloc.

Reply:
This (and other comments) apply in multiple places throughout, not going to make the duplicate comment in multiple places.

Reply:
Will make these changes after careful testing in future release.